metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,503 @@
+              Pat.Var.apply(Term.Name.apply(lastGeneratedName)),
+              nextQual
+            )
+          )
+          val newYield = Term.Name.apply(lastGeneratedName)
+          (newEnumerations, Some(newYield))
+        } else (existingForElements, Some(nextQual))
+    }
+  }
+
+  private def processMap(
+      elems: List[Enumerator],
+      maybeYieldTerm: Option[Term],
+      nameGenerator: MetalsNames,
+      valueName: String,
+      termSelectQual: Term
+  ): (List[Enumerator], Option[Term]) = {
+
+    termSelectQual match {
+      case qualTermApply: Term.Apply =>
+        extractChainedForYield(
+          Some(valueName),
+          maybeYieldTerm,
+          elems,
+          qualTermApply,
+          nameGenerator
+        )
+      case otherQual =>
+        (
+          Enumerator.Generator(
+            Pat.Var.apply(Term.Name.apply(valueName)),
+            otherQual
+          )
+            +: elems,
+          maybeYieldTerm
+        )
+
+    }
+
+  }
+
+  private def processFilter(
+      perhapsValueNameAndNextQual: Option[(String, Term)],
+      nameGenerator: MetalsNames,
+      perhapsLastName: Option[String],
+      isFilter: Boolean,
+      existingForElements: List[Enumerator],
+      currentYieldTerm: Option[Term],
+      termSelectQual: Term
+  ): (List[Enumerator], Option[Term]) = {
+
+    val result = for {
+      valueName <- perhapsValueNameAndNextQual.map(_._1)
+      nextCondition <- perhapsValueNameAndNextQual.map(_._2)
+    } yield {
+      val (elems, maybeYieldTerm): (List[Enumerator], Option[Term]) =
+        perhapsLastName match {
+          case Some(lastName) =>
+            (
+              Enumerator.Val.apply(
+                Pat.Var.apply(Term.Name.apply(lastName)),
+                Term.Name.apply(valueName)
+              ) +: existingForElements,
+              currentYieldTerm
+            )
+          case None =>
+            (existingForElements, Some(Term.Name.apply(valueName)))
+        }
+
+      termSelectQual match {
+        case qualTermApply: Term.Apply =>
+          extractChainedForYield(
+            Some(valueName),
+            maybeYieldTerm,
+            Enumerator.Guard.apply(
+              if (isFilter) nextCondition
+              else
+                Term.ApplyUnary.apply(Term.Name.apply("!"), nextCondition)
+            ) +: elems,
+            qualTermApply,
+            nameGenerator
+          )
+        case otherQual =>
+          (
+            Enumerator.Generator(
+              Pat.Var.apply(Term.Name.apply(valueName)),
+              otherQual
+            )
+              +: Enumerator.Guard.apply(
+                if (isFilter) nextCondition
+                else
+                  Term.ApplyUnary.apply(Term.Name.apply("!"), nextCondition)
+              ) +: elems,
+            maybeYieldTerm
+          )
+      }
+    }
+    result.getOrElse(List.empty, currentYieldTerm)
+  }
+
+  private def extractChainedForYield(
```

Comment:
Can you add short description explaining the behavior of this method in Scaladoc comment ?

Reply:
Sure, Rikito 💯